src/utils/Block.ts

```diff
@@ -0,0 +1,242 @@
+
+    this._registerEvents(eventBus);
+    eventBus.emit(Block.EVENTS.INIT);
+  }
+
+  _getChildren(propsAndChildren: any) {
+    const children: any = {};
+    const props: any = {};
+
+    Object.entries(propsAndChildren).forEach(([key, value]) => {
+      if (value instanceof Block) {
+        children[key] = value;
+      } else if (Array.isArray(value) && value.every((v) => (v instanceof Block))) {
+        children[key] = value;
+      } else {
+        props[key] = value;
+      }
+    });
+    return { children, props };
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-empty-function
+  protected initChildren() { }
+
+  _registerEvents(eventBus: EventBus) {
+    eventBus.on(Block.EVENTS.INIT, this.init.bind(this));
+    eventBus.on(Block.EVENTS.FLOW_CDM, this._componentDidMount.bind(this));
+    eventBus.on(Block.EVENTS.FLOW_CDU, this._componentDidUpdate.bind(this));
+    eventBus.on(Block.EVENTS.FLOW_RENDER, this._render.bind(this));
+  }
+
+  init() {
+    this.eventBus().emit(Block.EVENTS.FLOW_RENDER);
+  }
+
+  _componentDidMount() {
+    this.componentDidMount();
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-empty-function
+  protected componentDidMount() { }
+
+  dispatchComponentDidMount() {
+    this.eventBus().emit(Block.EVENTS.FLOW_CDM);
+  }
+
+  _componentDidUpdate(oldProps: any, newProps: any) {
+    if (this.componentDidUpdate(oldProps, newProps)) {
+      this.eventBus().emit(Block.EVENTS.FLOW_RENDER);
+    }
+  }
+
+  // eslint-disable-next-line @typescript-eslint/no-unused-vars
+  componentDidUpdate(oldProps: any, newProps: any) {
+    return true;
+  }
+
+  setProps = (nextProps: any) => {
+    if (!nextProps) {
+      return;
+    }
+
+    Object.assign(this.props, nextProps);
+  };
+
+  get element(): HTMLElement | null {
+    return this._element;
+  }
+
+  _render() {
+    const fragment = this.render();
+
+    const newElement = fragment.firstElementChild as HTMLElement;
+
+    if (this._element) {
+      this._removeEvents();
+      this._element.replaceWith(newElement);
+    }
+
+    this._element = newElement;
+
+    const elementForEvent = newElement.querySelector("[data-event]");
+    if (elementForEvent) {
+      this._elementForEvents = elementForEvent as HTMLElement;
+    }
+
+    this._removeEvents();
+
+    this._addEvents();
+  }
+
+  protected render(): DocumentFragment {
+    return new DocumentFragment();
+  }
+
+  getContent(): HTMLElement | null {
+    return this._element;
+  }
+
+  _makePropsProxy(props: any) {
+    const self = this;
+
+    return new Proxy(props, {
+      get(target: Record<string, unknown>, prop: string) {
+        const value = target[prop];
+        return typeof value === "function" ? value.bind(target) : value;
+      },
+
+      set(target: Record<string, unknown>, prop: string, value: unknown) {
+        const oldProps = { ...target };
+        target[prop] = value;
+        self.eventBus().emit(Block.EVENTS.FLOW_CDU, oldProps, target);
```

Comment:
Можно использовать стрелочные функции, тогда не придется прибегать к замыканию
```suggestion
      set: (target: Record<string, unknown>, prop: string, value: unknown) => {
        const oldProps = { ...target };
        target[prop] = value;
        this.eventBus().emit(Block.EVENTS.FLOW_CDU, oldProps, target);
```